apps/dispatch/dispatch-1.py

```diff
@@ -0,0 +1,275 @@
+					defaultValue = "",
+					allowEmptyString = True,
+					extensions = "gfr",
+					check = IECore.FileNameParameter.CheckType.MustExist,
+				),
+
+				IECore.BoolParameter(
+					name = "ignoreScriptLoadErrors",
+					description = "Causes errors which occur while loading the script "
+						"to be ignored. Not recommended.",
+					defaultValue = False,
+				),
+
+				IECore.BoolParameter(
+					name = "applyUserDefaults",
+					description = "Applies userDefault values to all nodes and plugs created by the app, "
+						"including the dispatchers. Note if a script is supplied, the nodes will be unaffected.",
+					defaultValue = False,
+				),
+
+				IECore.StringVectorParameter(
+					name = "nodes",
+					description = "The names of the nodes to dispatch. Note if a script is supplied, the nodes must "
+						"exist within the script. If no script is supplied, the nodes will be constructed on the fly "
+						"and added to a default script.",
+					defaultValue = IECore.StringVectorData( [] ),
+				),
+
+				IECore.StringParameter(
+					name = "dispatcher",
+					description = "The type of dispatcher to use for the dispatch. A new dispatcher "
+						"of this type will be created, with userDefaults applied.",
+					defaultValue = "",
+				),
+
+				IECore.StringVectorParameter(
+					name = "settings",
+					description = "The values to be set on the nodes, dispatcher, or context. Values "
+						"should be in the format -nodeA.plugA value -nodeA.plugB value -nodeB.plugC value "
+						"-dispatcher.plugD value -context.entry value",
+					defaultValue = IECore.StringVectorData( [] ),
+					userData = {
+						"parser" : {
+							"acceptFlags" : IECore.BoolData( True ),
+						},
+					},
+				),
+
+			]
+
+		)
+
+	def _run( self, args ) :
+
+		nodes = []
+		if not len( args["nodes"] ) :
+			IECore.msg( IECore.Msg.Level.Error, "gaffer dispatch", "No nodes were specified" )
+			return 1
+
+		script = Gaffer.ScriptNode()
+
+		if args["script"].value :
+			script["fileName"].setValue( os.path.abspath( args["script"].value ) )
+			try :
+				script.load( continueOnError = args["ignoreScriptLoadErrors"].value )
+			except Exception as exception :
+				IECore.msg( IECore.Msg.Level.Error, "gaffer dispatch : loading \"%s\"" % script["fileName"].getValue(), str( exception ) )
+				return 1
+
+		self.root()["scripts"].addChild( script )
+
+		for nodeName in args["nodes"] :
+			if args["script"].value :
+				node = script.descendant( nodeName )
+				if node is None :
+					IECore.msg( IECore.Msg.Level.Error, "gaffer dispatch", "\"%s\" does not exist." % nodeName )
+					return 1
+			else :
+				node = self.__create( nodeName )
+				if args["applyUserDefaults"].value :
+					Gaffer.NodeAlgo.applyUserDefaults( node )
+				script.addChild( node )
+			nodes.append( node )
+
+		dispatcherType = args["dispatcher"].value or GafferDispatch.Dispatcher.getDefaultDispatcherType()
+
+		if args["gui"].value :
```

Comment:
I originally wrote it (both this block and the constructor) the way you’re suggesting and then changed it to this when I realized we needed the drop down of dispatchers in the dialogue.

I don’t think it’s reasonable to require the client of DispatchDialogue to construct all the dispatchers upfront. That defeats the purpose of registering dispatchers, and in the majority of use cases we have today, it’s perfectly valid for the end user to choose any dispatcher.

It’s also not very friendly for the client of dispatch app. To make the gui work the command would be something like:

gaffer dispatch -gui -nodes SystemCommand -dispatchers Local Tractor

and we’d use the first dispatcher as the default. But then if you omitted the -gui what would we do? Surely not dispatch with both. Erroring for giving two things to a parameter with a plural name doesn’t seem right either...

One thing I realized I’m missing currently is applying the command line dispatcher settings to all dispatchers... currently the frameRange, for instance, can be out of sync, which probably isn’t desirable. I realize this would manifest itself as yet another dialogue constructor arg, but I think it’s probably the right thing to do.

Reply:
> I don’t think it’s reasonable to require the client of DispatchDialogue to construct all the dispatchers upfront.

It's totally reasonable; it's about giving control to the client code. For convenience it could have a default value which means "use all the registered dispatchers".

> in the majority of use cases we have today, it’s perfectly valid for the end user to choose any dispatcher.

Do you have a plan for the minority of cases?

> It’s also not very friendly for the client of dispatch app. To make the gui work the command would be something like:

> gaffer dispatch -gui -nodes SystemCommand -dispatchers Local Tractor

Again, the default would be "show 'em all". The command line argument could remain singular (`-dispatcher`), and constrain the UI to show only that dispatcher.

> I realize this would manifest itself as yet another dialogue constructor arg, but I think it’s probably the right thing to do.

The fact that it requires another dialogue constructor arg is a sign that the dialogue constructor isn't right. But I agree, we should apply the frame range to all the dispatchers. We should just do it outside the dialogue...

Reply:
> It's totally reasonable; it's about giving control to the client code.

@lento was actually in favor of your approach originally, but thinking of a future where the dispatchers are in the graph, and the tool author would drop down the ones that should be available, and expose a switch if they wanted to offer a choice. But that future is a ways off, and kind of implies we'd remove the entire Dispatcher tab, only exposing the plugs that were promoted intentionally.

And then I mentioned that if we added a 3rd dispatcher, he'd have to go updating all his tools to account for it, and he started seeing the value in my approach.

But lets say for the moment I agreed to yours...

> For convenience it could have a default value which means "use all the registered dispatchers".

Ok, sounds good.

> The command line argument could remain singular (-dispatcher), and constrain the UI to show only that dispatcher.

What about a command line invocation that wants to give users a choice of dispatchers?

>  But I agree, we should apply the frame range to all the dispatchers. We should just do it outside the dialogue...

So how would a client of the dialogue use both the default "use all the registered dispatchers" arg and also set the frame range on all of them to match? Seems like we still need the extra constructor arg to me.

---------------------

And back to my approach...

> it's about giving control to the client code.

Control comes with a maintenance burden.

> Do you have a plan for the minority of cases?

Trust the end user to be reasonable, rather than tying their hands to one dispatcher or the other.

Reply:
I feel like my last reply comes off more argumentative than I intended it. I’m willing to add the ‘client provides all the dispatchers’ feature if we can come up with an approach that also enables the ‘client provides minimal defaults and still gets all the other dispatchers’ workflow. That 2nd approach would have to work for both clients of the dialogue and clients of the commandline app to be useful at IE though.

Reply:
> What about a command line invocation that wants to give users a choice of dispatchers, but also specify a default dispatcher that doesn’t match the registered default dispatcher?

I guess that would require extra arguments. Is this something we need right now?

> So how would a client of the dialogue use both the default "use all the registered dispatchers" arg and also set the frame range on all of them to match? Seems like we still need the extra constructor arg to me.

Hmm, true. Very good point. But I'm afraid this doesn't push me towards wanting the dialogue to construct the dispatchers internally.

> Control comes with a maintenance burden.

Poor APIs come with a maintenance burden. If the dialogue constructs the dispatchers internally, we're going to have to tweak the API every time we find we need more control over that construction. We have perfectly good APIs for querying available dispatchers, creating them and setting their plugs. The permutations are many; I predict a mess if we try to hide them inside the dialogue and add more and more parameters as we find we need more control.

> Trust the end user to be reasonable, rather than tying their hands to one dispatcher or the other.

Personally I'm a fan of that approach, but we're always being asked to police problems via technology. And that was just a single example.

> I’m willing to add the ‘client provides all the dispatchers’ feature if we can come up with an approach that also enables the ‘client provides minimal defaults and still gets all the other dispatchers’ workflow.

Great! I propose the following :

- The DispatchDialogue constructor takes a list of already-constructed dispatchers as I've proposed. It doesn't have any friendly fallback to creating them all, since as you point out, that won't be very useful. This gives us flexibility and a simpler core API but with less convenience.
- We add a utility method somewhere which provides the convenience you're looking for. Perhaps a static `DispatchDialogue.createWithDefaultDispatchers()` method, or a `DispatcherAlgo::createAll()` method. I'll leave the details to you.

How does that sound?

Reply:
I’ll try that approach (probably tomorrow) and see how it feels.

>> What about a command line invocation that wants to give users a choice of dispatchers, but also specify a default dispatcher that doesn’t match the registered default dispatcher?
> I guess that would require extra arguments. Is this something we need right now?

Yes, we need this now. The default is Local, I’d like to keep it that way at IE, but I’d like quite a few of our command line invocations to start the gui with Farm selected and with Local available if the user sees fit to use it.

So what’s the plan for tackling this use case? It shouldn’t affect your design for DispatchDialogue, but will impact the app parameters in some way.

Reply:
It seems there's going to be a bit of conflict between the gui and non-gui flavours here whatever we do, because multiple dispatchers don't make sense for the non-gui flavour. I like the way you have it already, which is that a singular `-dispatcher` argument specifies the dispatcher used for the non-gui mode, and also the default dispatcher for gui mode. So maybe we just need an `-otherDispatchers` argument that is used to specify additional dispatchers to choose from in the gui mode? We could allow wildcards in that parameter so that `*` could be used to use all available dispatchers, without needing to know their names.

Reply:
Addressed in 57a006132da5bfa4a0ccfce3d710a6a17249a467, ea8aa5141ae3f7e44d51a06312da06d4ccefd8b1, 38ad464a325c8f111aedf2b673febd2986c5c908, 78549754d2099470c22159585b44606805cd7656